src/intelligence_layer/connectors/data/data.py

```diff
@@ -0,0 +1,245 @@
+import json
+import warnings
+from collections.abc import Iterator
+from http import HTTPStatus
+from typing import Any, ClassVar
+from urllib.parse import urlencode, urljoin
+
+import requests
+from requests import HTTPError
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+
+from intelligence_layer.connectors.data.exceptions import (
+    DataError,
+    DataExternalServiceUnavailable,
+    DataForbiddenError,
+    DataInternalError,
+    DataInvalidInput,
+    DataResourceNotFound,
+)
+from intelligence_layer.connectors.data.models import (
+    DataDataset,
+    DataRepository,
+    DataRepositoryCreate,
+    DatasetCreate,
+)
+
+
+class DataClient:
+    """Data Client class to interact with the Data Platform API.
+
+    Attributes:
+        headers: headers used in the request session
+    """
+
+    _status_code_to_exception: ClassVar[dict[int, type[DataError]]] = {
+        HTTPStatus.SERVICE_UNAVAILABLE: DataExternalServiceUnavailable,
+        HTTPStatus.NOT_FOUND: DataResourceNotFound,
+        HTTPStatus.UNPROCESSABLE_ENTITY: DataInvalidInput,
+        HTTPStatus.FORBIDDEN: DataForbiddenError,
+    }
+
+    def __init__(
+        self,
+        token: str | None,
+        base_data_platform_url: str = "http://localhost:8000",
+        session: requests.Session | None = None,
+    ) -> None:
+        """Initialize the Data Client.
+
+        Args:
+            token: Access token
+            base_data_platform_url: Base URL of the Data Platform API
+            session: Requests session.
```

Comment:
A more descriptive string would be good here. The type already tells you its a request session. This is the request session that will be used for all requests. It also has a default that should be in the docstring.

```
{{summaryPlaceholder}}

{{extendedSummaryPlaceholder}}
{{#parametersExist}}

Args:
{{#args}}
    {{var}}: {{descriptionPlaceholder}}
{{/args}}
{{#kwargs}}
    {{var}}: {{descriptionPlaceholder}}. Defaults to {{&default}}.
{{/kwargs}}
{{/parametersExist}}
{{#returnsExist}}

Returns:
{{#returns}}
    {{descriptionPlaceholder}}
{{/returns}}
{{/returnsExist}}
```
This is the template we are pretty much using for docstrings. Its a slight variation of google docstrings